packages/optimizer/lib/transformers/MinifyHtml.js

```diff
@@ -0,0 +1,172 @@
+    this.minifyNode(tree, this.opts, nodesToRemove);
+    for (const node of nodesToRemove) {
+      remove(node);
+    }
+  }
+
+  minifyNode(node, opts, nodesToRemove) {
+    if (node.type === 'text') {
+      this.minifyTextNode(node, opts, nodesToRemove);
+    } else if (node.type === 'comment') {
+      this.minifyCommentNode(node, opts, nodesToRemove);
+    } else if (node.tagName === 'script') {
+      this.minifyScriptNode(node, opts);
+    }
+    // update options based on the current node
+    const childOpts = Object.assign({}, opts);
+    if (opts.canCollapseWhitespace && !this.canCollapseWhitespace(node.tagName)) {
+      childOpts.canCollapseWhitespace = false;
+    }
+    if (opts.canTrimWhitespace && !this.canTrimWhitespace(node.tagName)) {
+      childOpts.canTrimWhitespace = false;
+    }
+    // minify all child nodes
+    for (const child of node.children || []) {
+      this.minifyNode(child, childOpts, nodesToRemove);
+    }
+  }
+
+  minifyTextNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.collapseWhitespace) {
+      return;
+    }
+    if (opts.canCollapseWhitespace) {
+      node.data = node.data.replace(WHITESPACE_REGEX, ' ');
+    }
+    if (opts.canTrimWhitespace) {
+      node.data = node.data.trim();
+    }
+    // remove empty nodes
+    if (node.data.length === 0) {
+      nodesToRemove.push(node);
+    }
+  }
+
+  minifyCommentNode(node, opts, nodesToRemove) {
+    if (!node.data || !opts.removeComments) {
+      return;
+    }
+    if (opts.commentIgnorePattern.test(node.data)) {
+      return;
+    }
+    nodesToRemove.push(node);
+  }
+
+  minifyScriptNode(node, opts) {
+    const isJson = this.isJson(node);
+    const isAmpScript = this.isInlineAmpScript(node);
+    for (const child of node.children || []) {
+      if (!child.data) {
+        continue;
+      }
+      if (isJson && opts.minifyJSON) {
+        this.minifyJson(child);
+      } else if (isAmpScript && opts.minifyAmpScript) {
+        this.minifyAmpScript(child);
+      }
+    }
+  }
+
+  minifyAmpScript(child) {
+    const result = Terser.minify(child.data);
+    if (result.error) {
+      this.log.warn(
+          'Could not minify inline amp-script',
+          child.data,
+          `${result.error.name}: ${result.error.message}`,
+      );
+      return;
+    }
+    child.data = result.code;
+  }
+
+  minifyJson(child) {
+    try {
+      child.data = JSON.stringify(JSON.parse(child.data), null, '');
+    } catch (e) {
+      // log invalid JSON, but don't fail
+      this.log.warn('Invalid JSON', child.data);
+    }
+  }
+
+  isInlineAmpScript(node) {
+    return node.attribs &&
+      node.attribs.type === 'text/plain' &&
+      node.attribs.target === 'amp-script';
+  }
+
+  isJson(node) {
+    return node.attribs &&
+      (node.attribs.type === 'application/json' || node.attribs.type === 'application/ld+json');
+  }
+
+  canCollapseWhitespace(tagName) {
+    return !/^(?:script|style|pre|textarea)$/.test(tagName);
```

Comment:
Not sure how important performance is here, but as this runs over every loop, it might be worthwhile to test using direct string comparisons here instead of a regex.

Reply:
just tested this, non-regex is 25% faster. Addressed in https://github.com/ampproject/amp-toolbox/pull/591/commits/62c343e5685670d67a5fd3e0ff348bf0867d3c89